File: rust_exercises/leetcode/Problem743/Solution.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        Map<Integer, List<int[]>> edges = new HashMap<>();
        Map<Integer, Integer> delay = new HashMap<>();
        
        for (int[] time: times) {
            if (!edges.containsKey(time[0])) {
                edges.put(time[0], new ArrayList<>());
            }
            edges.get(time[0]).add(new int[]{time[1], time[2]});
        }
        delay.put(k, 0);
        
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(k, 0));
        
        while (!pq.isEmpty()) {
            Node closest = pq.poll();
            if (!edges.containsKey(closest.id)) {
                continue;
            }
            for (int[] pair: edges.get(closest.id)) {
                int dest = pair[0];
                int time = closest.time + pair[1];
                if (!delay.containsKey(dest)) {
                    delay.put(dest, time);
                    pq.offer(new Node(dest, time));
                } else {
                    if (time < delay.get(dest)) {
                        delay.replace(dest, time);
                        pq.offer(new Node(dest, time));
                    }
                }
            }
        }
        
        if (delay.size() < n) {
            return -1;
        }
        
        int result = 0;
        for (Map.Entry<Integer, Integer> entry: delay.entrySet()) {
            result = Math.max(result, entry.getValue());
        }
        
        return result;
    }
}

class Node implements Comparable<Node> {
    int id;
    int time;
    
    public Node(int id, int time) {
        this.id = id;
        this.time = time;
    }
    
    @Override
    public int compareTo(Node n) {
        return this.time - n.time;
    }
}

File: rust_exercises/leetcode/Problem743/solution.rs
use std::collections::HashMap;
use std::collections::BinaryHeap;
use std::cmp::Ordering;
use std::cmp::Reverse;
use std::cmp::max;

impl Solution {
    pub fn network_delay_time(times: Vec<Vec<i32>>, n: i32, k: i32) -> i32 {
        let mut edges = HashMap::new();
        for edge in times.into_iter() {
            let src = edge[0];
            let dst = edge[1];
            let time = edge[2];
            if !edges.contains_key(&src) {
                edges.insert(src, vec![]);
            }
            edges.get_mut(&src).unwrap().push((dst, time));
        }
        
        let mut delay = HashMap::new();
        let mut pq = BinaryHeap::new();
        
        pq.push(Node {
            id: k,
            time: 0,
        });
        
        while !pq.is_empty() {
            let curr = pq.pop().unwrap();
            if delay.contains_key(&curr.id) {
                continue;
            }
            delay.insert(curr.id, curr.time);
            if !edges.contains_key(&curr.id) {
                continue;
            }
            
            for &(dst, time) in edges.get(&curr.id).unwrap().iter() {
                if !delay.contains_key(&dst) {
                    pq.push(Node {
                        id: dst,
                        time: curr.time + time,
                    });
                }
            }
        }
        
        if (delay.len() as i32) < n {
            return -1;
        }
        
        let mut result = 0;
        for (&_, &v) in delay.iter() {
            result = max(result, v);
        }
        result
    }
}

#[derive(Eq)]
struct Node {
    id: i32,
    time: i32,
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.time != other.time {
            Reverse(self.time).cmp(&Reverse(other.time))
        } else {
            self.id.cmp(&other.time)
        }
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl PartialEq for Node {
    fn eq(&self, other: &Self) -> bool {
        self.time == other.time && self.id == other.id
    }
}
